drivers/flash/flash_realtek_rts5912.c

```diff
@@ -0,0 +1,826 @@
+		return 0;
+	}
+
+	if ((offset % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+
+	if ((len % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+
+	ret = check_boundary(offset, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	for (; len > 0; len -= FLASH_ERASE_BLK_SZ) {
+		ret = flash_erase_sector(dev, offset);
+		if (ret < 0) {
+			LOG_ERR("erase @0x%08lx fail", offset);
+		}
+		offset += FLASH_ERASE_BLK_SZ;
+	}
+
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int flash_rts5912_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	int ret;
+	unsigned int key;
+
+	if (len == 0) {
+		return 0;
+	}
+
+	ret = check_boundary(offset, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+	key = irq_lock();
+	ret = flash_program_page(dev, offset, data, len);
+	irq_unlock(key);
+	k_sem_give(&dev_data->sem);
+
+	return ret;
+}
+
+static int flash_rts5912_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	int ret;
+
+	if (len == 0) {
+		return 0;
+	}
+
+	ret = check_boundary(offset, len);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+	ret = flash_normal_read(dev, SPI_NOR_CMD_READ, offset, data, len);
+	k_sem_give(&dev_data->sem);
+
+	return ret;
+}
+
+static const struct flash_parameters *flash_rts5912_get_parameters(const struct device *dev)
+{
+	const struct flash_rts5912_dev_config *config = dev->config;
+
+	return &config->flash_rts5912_parameters;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static const struct flash_pages_layout dev_layout = {
+	.pages_count =
+		DT_REG_SIZE(SOC_NV_FLASH_NODE) / DT_PROP(SOC_NV_FLASH_NODE, erase_block_size),
+	.pages_size = DT_PROP(SOC_NV_FLASH_NODE, erase_block_size),
+};
+
+static void flash_rts5912_pages_layout(const struct device *dev,
+				       const struct flash_pages_layout **layout,
+				       size_t *layout_size)
+{
+	*layout = &dev_layout;
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+static int flash_rts5912_ex_op(const struct device *dev, uint16_t opcode, const uintptr_t in,
+			       void *out)
+{
+	struct flash_rts5912_dev_data *dev_data = dev->data;
+	int ret = -EINVAL;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	switch (opcode) {
+	case FLASH_RTS5912_EX_OP_WR_ENABLE:
+		ret = flash_write_enable(dev);
+		break;
+	case FLASH_RTS5912_EX_OP_WR_DISABLE:
+		ret = flash_write_disable(dev);
+		break;
+	case FLASH_RTS5912_EX_OP_WR_SR:
+		ret = flash_write_status_reg(dev, (uint8_t *)out, 1);
+		break;
+	case FLASH_RTS5912_EX_OP_WR_SR2:
+		ret = flash_write_status_reg2(dev, (uint8_t *)out, 1);
+		break;
+	case FLASH_RTS5912_EX_OP_RD_SR:
+		ret = flash_read_sr(dev, (uint8_t *)in);
+		break;
+	case FLASH_RTS5912_EX_OP_RD_SR2:
+		ret = flash_read_sr2(dev, (uint8_t *)in);
+		break;
```

Comment:
fwiw you have in and out backwards here, I believe (came across this as I am currently adding doxygen docs for the rts5912_flash_api_ex.h header and this is really confusing IMO